src/tracks/Pattern.cpp

```diff
@@ -213,7 +213,7 @@ Note * Pattern::addNote( const Note & _new_note, const bool _quant_pos )
 	}
 
 	instrumentTrack()->lock();
-	if( m_notes.size() == 0 || m_notes.back()->pos() <= new_note->pos() )
+	if( m_notes.size() == 0 || m_notes.back()->pos() < new_note->pos() )
```

Comment:
As far as I understand the code, all lines 216 to 245 do is insert the new note at the right position (sorted by the criteria in `lessThan`). If so, we can reduce this to a single line:
```cpp
m_notes.insert(std::upper_bound(m_notes.begin(), m_notes.end(), new_note, Note:lessThan), new_note);
```
This reduces code complexity, improves performance by doing a binary search, and removes the duplicate compare code by using the `lessThan` method.

Reply:
> As far as I understand the code, all lines 216 to 245 do is insert the new note at the right position (sorted by the criteria in lessThan).

Yes indeed. Oneliner, nice but I can't make it work.

Computer says no. 8)

```cpp
/usr/include/c++/4.8/bits/stl_algo.h: In instantiation of ‘_FIter std::upper_bound(_FIter, _FIter, const _Tp&, _Compare) [with _FIter = Note**; _Tp = Note*; _Compare = bool (*)(Note*&, Note*&)]’:
/home/zonkmachine/builds/lmms/lmms/src/tracks/Pattern.cpp:216:90:   required from here
/usr/include/c++/4.8/bits/stl_algo.h:2543:31: error: invalid initialization of reference of type ‘Note*&’ from expression of type ‘Note* const’
    if (__comp(__val, *__middle))
                               ^
```

Reply:
Ah, you'll have to change `Note::lessThan` to take `const` pointers.

Reply:
Perfect, thanks!